### PR TITLE
Added fr-ch localisation

### PR DESCRIPTION
**Pull request by mrzhdev**
Added Swiss french localisation with translations

## Description
Added Swiss French localization with translations.
Used Zeta Resource Editor and related files are available upon request.
Probably the naming conventions are not perfect, as I am not an accountant either.

## Related issue
None.

### CODE REVIEW

**Comment by mrzhdev** [comment]
Hi Lucas,

The main reason that I have used Swiss French localization is the use of
CHF as the currency, which is set directly in the resource file. It might
be a good idea to have currency in the settings, perhaps even a decimal
separator (Swiss decimal separator is comma and I use dot).  What do you
think?

My aim is to use your software for my accounting as a freelancer and extend
it for my personal use. Anyway, I’ll be glad to contribute, share my
development, and maintain the French localization.

++

Mirza

Le mer. 9 nov. 2022 à 09:30, Lukas Grützmacher ***@***.***> a
écrit :

**Comment by lg2de** [comment]
The currency should be configurable in the project. It must not be derived from the culture settings.
This an issue which should be tracked separately.
It it is (fully) part of the resources (I've not checked that), then this is a bug.

The decimal separator should come from culture. This information is already in and should not be configured.
To reduce the effort and maximize the effect we can
1. Rename `Resources.fr-ch.resx` to `Resources.fr.resx` and update if needed.
2. Add "fr" and "fr-ch" to the culture selector. If "fr-ch" is selected the "resource culture fr" should be used together with the "ui culture fr-ch" (to get the dot as decimal separator).

**Comment by lg2de** [comment]
With #127 we have already an issue for the currency configuration feature.

**Comment by mrzhdev** [comment]
In resources you have the field "Label_ValueWithCurrency" with "Value (€)"
in english.
It's the only reason why I have to go with "fr-ch"instead of "fr".

**Comment by lg2de** [comment]
Yes, this was my fault - due to simplification for my so-far-use-cases.
Please just switch to "fr" only. You can also leave "CHF" for now.
I'll take care configuration of currency in the project incl. updating this resource.

File: src/SimpleAccounting/Properties/Resources.fr-ch.resx [hunk 1]
```diff
@@ -0,0 +1,570 @@
...
+    <value>Texte de journalisation</value>
+  </data>
+	<data name="Tooltip_SplitBookingValue" xml:space="preserve">
+    <value>Diviser la valeur de la journalisation</value>
+  </data>
+	<data name="Word_BookingNumber" xml:space="preserve">
+    <value>Numéro de journalisation</value>
+  </data>
+	<data name="Label_ValueWithCurrency" xml:space="preserve">
+    <value>Valeur (CHF)</value>
```
Comment (lg2de) [inline src/SimpleAccounting/Properties/Resources.fr-ch.resx:219]:
```suggestion
    <value>Valeur ({0})</value>
```

File: src/SimpleAccounting/Properties/Resources.fr-ch.resx [hunk 2]
```diff
@@ -0,0 +1,570 @@
...
+	<data name="Word_BookingText" xml:space="preserve">
+    <value>Texte de journalisation</value>
+  </data>
+	<data name="Tooltip_SplitBookingValue" xml:space="preserve">
+    <value>Diviser la valeur de la journalisation</value>
+  </data>
+	<data name="Word_BookingNumber" xml:space="preserve">
+    <value>Numéro de journalisation</value>
+  </data>
+	<data name="Label_ValueWithCurrency" xml:space="preserve">
```
Comment (lg2de) [inline src/SimpleAccounting/Properties/Resources.fr-ch.resx:218]:
```suggestion
	<data name="Label_ValueWithCurrencyX" xml:space="preserve">
```

File: src/SimpleAccounting/Properties/Resources.fr-ch.resx [hunk 3]
```diff
@@ -0,0 +1,570 @@
...
+	<data name="Word_Expression" xml:space="preserve">
+    <value>Expression</value>
+  </data>
+	<data name="AccountConfiguration_SourceDescription" xml:space="preserve">
+    <value>Nom de la colonne dans le fichier d'importation (CSV)</value>
+  </data>
+	<data name="AccountConfiguration_IgnorePatternDescription" xml:space="preserve">
+    <value>Expression régulière utilisée pour supprimer des sous-chaînes</value>
+  </data>
+</root>
```
Comment (lg2de) [inline src/SimpleAccounting/Properties/Resources.fr-ch.resx:570]:
```suggestion
  <data name="Word_Currency" xml:space="preserve">
    <value>Currency</value>
  </data>
  <data name="Menu_Project_Settings_ProjectOptions" xml:space="preserve">
    <value>Options du projet</value>
  </data>
  <data name="Header_ProjectOptions" xml:space="preserve">
    <value>Modifier les options du projet</value>
  </data>
</root>
```

File: src/SimpleAccounting/Presentation/IMenuViewModel.cs [hunk 4]
```diff
@@ -22,6 +22,7 @@ internal interface IMenuViewModel : INotifyPropertyChanged
         ICommand SwitchCultureCommand { get; }
         bool IsGermanCulture { get; }
         bool IsEnglishCulture { get; }
+        bool IsSwissFrenchCulture { get; }
```
Comment (lg2de) [inline src/SimpleAccounting/Presentation/IMenuViewModel.cs:25]:
```suggestion
        bool IsFrenchCulture { get; }
        bool IsSwissFrenchCulture { get; }
```

File: src/SimpleAccounting/Presentation/MenuViewModel.cs [hunk 5]
```diff
@@ -67,6 +67,7 @@ internal class MenuViewModel : Screen, IMenuViewModel
                 this.projectData.Settings.Culture = cultureName.ToString();
                 this.NotifyOfPropertyChange(nameof(this.IsGermanCulture));
                 this.NotifyOfPropertyChange(nameof(this.IsEnglishCulture));
+                this.NotifyOfPropertyChange(nameof(this.IsSwissFrenchCulture));
```
Comment (lg2de) [inline src/SimpleAccounting/Presentation/MenuViewModel.cs:70]:
```suggestion
                this.NotifyOfPropertyChange(nameof(this.IsFrenchCulture));
                this.NotifyOfPropertyChange(nameof(this.IsSwissFrenchCulture));
```

File: src/SimpleAccounting/Presentation/MenuViewModel.cs [hunk 6]
```diff
@@ -76,6 +77,7 @@ internal class MenuViewModel : Screen, IMenuViewModel
 
         public bool IsGermanCulture => this.projectData.Settings.Culture == "de";
         public bool IsEnglishCulture => this.projectData.Settings.Culture == "en";
+        public bool IsSwissFrenchCulture => this.projectData.Settings.Culture == "fr-CH";
```
Comment (lg2de) [inline src/SimpleAccounting/Presentation/MenuViewModel.cs:80]:
```suggestion
        public bool IsFrenchCulture => this.projectData.Settings.Culture == "fr";
        public bool IsSwissFrenchCulture => this.projectData.Settings.Culture == "fr-CH";
```

File: src/SimpleAccounting/Presentation/ShellView.xaml [hunk 7]
```diff
@@ -52,6 +52,8 @@
                       IsCheckable="True" IsChecked="{Binding Menu.IsGermanCulture, Mode=OneWay}" />
             <MenuItem Header="English" Command="{Binding Menu.SwitchCultureCommand}" CommandParameter="en"
                       IsCheckable="True" IsChecked="{Binding Menu.IsEnglishCulture, Mode=OneWay}" />
+            <MenuItem Header="SwissFrench" Command="{Binding Menu.SwitchCultureCommand}" CommandParameter="fr-CH"
```
Comment (lg2de) [inline src/SimpleAccounting/Presentation/ShellView.xaml:55]:
```suggestion
            <MenuItem Header="Français" Command="{Binding Menu.SwitchCultureCommand}" CommandParameter="fr"
                      IsCheckable="True" IsChecked="{Binding Menu.IsFrenchCulture, Mode=OneWay}" />
            <MenuItem Header="Français suisse" Command="{Binding Menu.SwitchCultureCommand}" CommandParameter="fr-CH"
```

File: tests/SimpleAccounting.UnitTests/Presentation/MenuViewModelTests.cs [hunk 8]
```diff
@@ -77,6 +77,7 @@ public void SwitchCultureCommand_DummyLanguage_MessageBoxShownAndConfigurationUp
                 icon: MessageBoxImage.Information);
             sut.IsGermanCulture.Should().BeFalse();
             sut.IsEnglishCulture.Should().BeFalse();
+            sut.IsSwissFrenchCulture.Should().BeFalse();
```
Comment (lg2de) [inline tests/SimpleAccounting.UnitTests/Presentation/MenuViewModelTests.cs:80]:
```suggestion
            sut.IsFrenchCulture.Should().BeFalse();
            sut.IsSwissFrenchCulture.Should().BeFalse();
```

File: src/SimpleAccounting/Properties/Resources.fr-ch.resx [hunk 9]
```diff
@@ -0,0 +1,570 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (lg2de) [inline src/SimpleAccounting/Properties/Resources.fr-ch.resx:1]: Please rename the file to `Resources.fr.resx`

**Review by lg2de — CHANGES_REQUESTED** [review]

**Comment by mrzhdev** [comment]
Hi,

I have made a new push.

++

Mirza

**Comment by lg2de** [comment]
> I have made a new push.

Why you did not took my suggestions?
1. We can use both, french and french swiss, isn't it? Or are the differences that much?
2. You must update the resources according to my recent feature update to CONFIGURE the currency.

**Comment by mrzhdev** [comment]
Hi,

1.I haven't noticed that you have changed your mind since 11 november. Also
my knowledge of github is not perfect.
2.I'll see what is happening...

**Comment by mrzhdev** [comment]
I would suggest having only French and changing App.xaml.cs (see below).
Number and DateTime formats are copied from windows settings.

What do you think?


private void ApplicationStartup(object sender, StartupEventArgs e)
        {

            var settings = Settings.Default;
            if (!string.IsNullOrEmpty(settings.Culture))
            {
                var culture =
(CultureInfo)CultureInfo.GetCultureInfo(settings.Culture).Clone();
                //copy current user settings
                culture.NumberFormat =
CultureInfo.CurrentCulture.NumberFormat;
                culture.DateTimeFormat =
CultureInfo.CurrentCulture.DateTimeFormat;
                //apply culture to current thread
                CultureInfo.CurrentCulture = CultureInfo.CurrentUICulture =
culture;

            }
            else
            {
                // set control culture according to system culture
                //
https://stackoverflow.com/questions/4041197/how-to-set-and-change-the-culture-in-wpf
                FrameworkElement.LanguageProperty.OverrideMetadata(
                    typeof(FrameworkElement),
                    new
FrameworkPropertyMetadata(XmlLanguage.GetLanguage(CultureInfo.CurrentCulture.IetfLanguageTag)));
            }

        }

**Comment by lg2de** [comment]
I guess we should start again from the beginning...
You want to have French (swiss) localization and French swiss currency printing.
[I learned meanwhile](https://learn.microsoft.com/en-us/globalization/locale/currency-formatting) that "French swiss" is the one and only area where we have a difference in number formatting (.) and currency formatting (,), correct.
This means I/we need not only to change the available cultures or resources, but also all occurrences for currency formatting. This is quite big effort. But I think it should be correct.

I introduced manual culture selection because I faced some situations where automatic detection (system culture) was not the desired result.
But when I start selecting the culture I would like be consistent. So I do not like to decouple these settings.

I think with your initial commit and my suggestions we are quite close to the correct solution for you. Then we "only" need to change "all the currency formatting". (This I would like to handle in a different PR.)

If you still want to contribute this feature, please
1. create a new PR with your initial commit (I recommend to use a branch in your repository too)
2. rebase it on my main
3. Apply the suggesions above (just provide French - without "swiss")

If your computer is configured to "French swiss" SimpleAccounting with the "<System>" selection should automatically show the "French" localization and use the French swiss formatting.

**Comment by mrzhdev** [comment]
I have just found out appropriate WPF solution:

·       Adding to control ->
xmlns:globalization="clr-namespace:System.Globalization;assembly=mscorlib"

·       Adding to binding -> , ConverterCulture={x:Static
globalization:CultureInfo.CurrentCulture}

It’s working in designtime and runtime

https://www.codeproject.com/Tips/1004834/Binding-with-Respect-to-CurrentCulture

Exemple (ShellView.xaml line 157):

            <DataGrid.Columns>

              <DataGridTextColumn Header="{x:Static
properties:Resources.Word_Date}" Width="60"

                                  Binding="{Binding Date,
StringFormat=\{0:d\}, ConverterCulture={x:Static
globalization:CultureInfo.CurrentCulture}}" />

              <DataGridTextColumn Header="{x:Static
properties:Resources.Word_BookingNumber_Short}" Width="75"

                                  Binding="{Binding Identifier}" />

              <DataGridTextColumn Header="{x:Static
properties:Resources.Word_BookingText}" Width="200"

                                  Binding="{Binding Text}" />

              <DataGridTextColumn Header="{x:Static
properties:Resources.Word_Value}" Width="70"

                                  Binding="{Binding Value,
StringFormat=\{0:0.00\}, ConverterCulture={x:Static
globalization:CultureInfo.CurrentCulture}}"

**Comment by lg2de** [comment]
I don't want to miss your localization.
Therefore I created a new PR based on your work (cherry-pick).
Please review #213.

Further I understand what you want to change to archive correct visualization of overwritten regional settings.
A PR is welcome. Otherwise we should track this as an issue and I'll take care later.

Please note that - as stated before - we should also switch from "number formatting" to "currency formatting" to be fully correct in "swiss french" localization.

**Comment by mrzhdev** [comment]
Nicely done.

I have taken a look, and everything is fine.

On my side, I'll do my best to submit a consistent PR aligned with coding
style.

I'll keep following what is happening on the project and keep the French
translations updated.

**Pull request closed without merging**